include/unifex/unstoppable_baton.hpp

```diff
@@ -0,0 +1,165 @@
+
+  explicit _sender(unstoppable_baton& baton) noexcept
+    : baton_(&baton) {}
+
+  template (typename Receiver)
+    (requires receiver_of<Receiver>)
+  operation<std::decay_t<Receiver>> connect(Receiver&& r) const noexcept(
+      std::is_nothrow_constructible_v<std::decay_t<Receiver>, Receiver>) {
+    return operation<std::decay_t<Receiver>>{*baton_, (Receiver&&)r};
+  }
+
+ private:
+  unstoppable_baton* baton_;
+};
+
+struct unstoppable_baton {
+  unstoppable_baton() noexcept
+    : unstoppable_baton(false) {}
+
+  explicit unstoppable_baton(bool startSignalled) noexcept
+    : state_(startSignalled ? this : nullptr) {}
+
+  void post() noexcept;
+
+  bool ready() const noexcept {
+    return state_.load(std::memory_order_acquire) ==
+        static_cast<const void*>(this);
+  }
+
+  void reset() noexcept {
+    // transition from signalled (i.e. state_ == this) to not-signalled
+    // (i.e. state_ == nullptr).
+    void* oldState = this;
+
+    // We can ignore the the result.  We're using _strong so it won't fail
+    // spuriously; if it fails, it means it wasn't previously in the signalled
+    // state so resetting is a no-op.
+    (void)state_.compare_exchange_strong(
+        oldState, nullptr, std::memory_order_acq_rel, std::memory_order_relaxed);
+  }
+
+  [[nodiscard]] _sender wait() noexcept {
+    return _sender{*this};
+  }
+
+ private:
+  std::atomic<void*> state_{};
+
+  friend struct _op_base;
+
+  static void start_or_wait(_op_base& op, unstoppable_baton& baton) noexcept;
+};
+
+struct _op_base {
+  // note: next_ is intentionally left indeterminate until the operation is
+  //       pushed on the baton's stack of waiting operations
+  //
+  // note: next_ is the first member so that list operations don't have to
+  //       offset into this struct; hopefully that leads to smaller code
+  _op_base* next_;
+  unstoppable_baton* baton_;
+  void (*setValue_)(_op_base*) noexcept;
+
+  explicit _op_base(unstoppable_baton& baton, void (*setValue)(_op_base*) noexcept)
+    : baton_(&baton), setValue_(setValue) {}
+
+  ~_op_base() = default;
+
+  _op_base(_op_base&&) = delete;
+  _op_base& operator=(_op_base&&) = delete;
+
+  void set_value() noexcept {
+    setValue_(this);
+  }
+
+  void start() noexcept {
+    unstoppable_baton::start_or_wait(*this, *baton_);
+  }
+};
+
+template <typename Receiver>
+struct _operation<Receiver>::type : private _op_base {
+  explicit type(unstoppable_baton& baton, Receiver r)
+      noexcept(std::is_nothrow_move_constructible_v<Receiver>)
+    : _op_base(baton, &set_value_impl),
+      receiver_(std::move(r)) {}
+
+  ~type() = default;
+
+  type(type&&) = delete;
+  type& operator=(type&&) = delete;
+
+  using _op_base::start;
+
+ private:
+  UNIFEX_NO_UNIQUE_ADDRESS Receiver receiver_;
+
+  static void set_value_impl(_op_base* base) noexcept {
+    auto self = static_cast<type*>(base);
+
+    auto op = connect(schedule(), std::move(self->receiver_));
```

Comment:
Composing `schedule()` in like this means that, if `receiver_` has a stoppable stop token, then the scheduler may respond to it and `set_done`.  Given the baton is unstoppable, this feels awkwardly inconsistent.  Should I also compose in a `with_query_value` that ensures the scheduler only receives an unstoppable token?

Reply:
That would be my preference. @lewissbaker ?

Reply:
I think that semantic seems fine for now.

It's probably worth encapsulating the logic for an adapter that prevents stop-requests from propagating to a child sender into an `unstoppable(s)` algorithm. It would just wrap the parent receiver in something that hooks `get_stop_token()` to return an `unstoppable_token`.